### ISSUE TITLE
LZ4DecoderStream followed by MD5 CryptoStream in read mode is insanely slow

### ISSUE THREAD

**Issue by hooyao** [MiloszKrajewski/K4os.Compression.LZ4#32]
It's barely an issue, it's quite interesting though.
```
using (var hasher = MD5.Create())
{
    using (SomeStream ss = new SomeStream())) //Not important, assume this stream is fast, you can use a filestream.
    using (LZ4DecoderStream lz4ds = LZ4Stream.Decode(ss))
    using (CryptoStream md5HashStream = new CryptoStream(lz4ds, hasher, CryptoStreamMode.Read))
    using (FileStream writeFs = new FileStream(destPath, FileMode.OpenOrCreate, FileAccess.ReadWrite, FileShare.ReadWrite,  1024 * 1024))
    {
        md5HashStream.CopyTo(writeFs, 1024 * 1024);
    }
}
```
The above code is extremely slow, approximately 14MB/s on 8700K with nvme ssd.
Simply use CryptoStream in write mode as following
```
using (var hasher = MD5.Create())
{
    using (SomeStream ss = new SomeStream()))
    using (LZ4DecoderStream lz4ds = LZ4Stream.Decode(ss))
    using (FileStream writeFs = new FileStream(destPath, FileMode.OpenOrCreate,
                                FileAccess.ReadWrite, FileShare.ReadWrite,1024 * 1024))
    using (CryptoStream md5HashStream = new CryptoStream(writeFs, hasher, CryptoStreamMode.Write))   
    {
        lz4ds.CopyTo(md5HashStream,  1024 * 1024);
    }
}
```
The stream can reach 400MB/s on the same hardware.
I used visual studio and dottrace to profile the code, none of them worked, they reported 80% the cpu time was spent with `System Code`. The dottrace snapshot is attached here
[dottrace.zip](https://github.com/MiloszKrajewski/K4os.Compression.LZ4/files/4593657/dottrace.zip)

My wild guess is the low-level optimization used in K4os.Compression.LZ4 conflicts with the MD5 implementation, CPU cache thing? Hope you can find the reason.

**Comment by MiloszKrajewski** [comment]
Hi, It may take a while before I can investigate this, but I was just thinking: did you try to isolate them with BufferedStream between?

**Comment by hooyao** [comment]
@MiloszKrajewski Yes, adding a Bufferedstream between the LZ4DecoderStream and CryptoStream didn't work, even with a super large buffer, like 300MB, the streaming speed got stuck at 14MB/s.

**Comment by MiloszKrajewski** [comment]
Good news:
I know what's the problem is and how to "fix" it. I say "fix" as there is no bug, but apparently `CryptoStream` does not like when underlying stream does not return whole blocks at once.

Bad news:
Not today.

**Comment by hooyao** [comment]
@MiloszKrajewski  Thank you

**Comment by MiloszKrajewski** [comment]
Most likely fixed in 1.2.2-beta

**Comment by hooyao** [comment]
Thank you @MiloszKrajewski , the new release works.
